composer/index.ts

```diff
@@ -0,0 +1,168 @@
+        public total_instances: number
+    ) {}
+}
+
+class Composition {
+    constructor(public global: Global, public runs: Run[]) {}
+}
+
+async function main() {
+    sqlite3.verbose();
+
+    // Call sqlite to process the csv resource files and generate a database.
+    // We call the sqlite process instead of doing it here cause
+    // the dot commands are interpreted by the sqlite cli tool not sqlite itself,
+    // and it is a lot faster parsing the csv's.
+    const { stdout, stderr } = await execFile("sqlite3", [
+        DB,
+        ".mode csv",
+        ".import transports.csv transports",
+        ".import muxers.csv muxers",
+    ]);
+    if (stderr != "") {
+        throw new Error(`Could not parse csv resources: ${stderr}`);
+    }
+
+    const db = await open({
+        filename: DB,
+        driver: sqlite3.Database,
+    });
+
+    // Generate the testing combinations by SELECT'ing from both transports
+    // and muxers tables the distinct combinations where the transport and the muxer
+    // of the different libp2p implementations match.
+    const queryResults =
+        await db.all(`SELECT DISTINCT a.id as id1, b.id as id2, a.transport, ma.muxer
+                     FROM transports a, transports b, muxers ma, muxers mb
+                     WHERE a.id != b.id
+                     AND a.transport == b.transport
+                     AND a.id == ma.id
+                     AND b.id == mb.id
+                     AND ma.muxer == mb.muxer;`);
+    await db.close();
+
+    let global = new Global(
+        "multidimensional-testing",
+        "multidimensional",
+        argv.total_instances
+    );
+    let composition = new Composition(global, []);
+
+    for (let row of queryResults) {
+        // Instance count is hardcoded to 1 for now.
+        let instance = new Instance(1);
+
+        let build_args1 = new BuildArgs(
+            row.transport,
+            row.muxer,
+            row.id1,
+            argv.gitRef,
+            argv.gitTarget
+        );
+        let build_config1 = new BuildConfig(build_args1);
+        let group1 = new Group(row.id1, BUILDER, instance, build_config1);
+
+        let build_args2 = new BuildArgs(
+            row.transport,
+            row.muxer,
+            row.id2,
+            argv.gitRef,
+            argv.gitTarget
+        );
+        let build_config2 = new BuildConfig(build_args2);
+        let group2 = new Group(row.id2, BUILDER, instance, build_config2);
+
+        let run = new Run(
+            `${row.id1} x ${row.id2} x ${row.transport} x ${row.muxer}`,
+            [group1, group2]
+        );
+
+        composition.runs.push(run);
+    }
+
+    // Write the TOML file and remove the database file to avoid corrupting
+    // future runs.
+    await writeFile("composition.toml", toml.stringify(composition as any));
+    await unlink(DB);
+}
+
+main()
+    .then(() => process.exit(0))
+    .catch((err) => {
+        console.error(err);
+        process.exit(1);
```

Comment:
(another thread)

> Wdyt of using just docker:generic to simplify the process @laurentsenta @mxinden?

If you want to tackle this question as part of this PR:

Initially, we planned on:

- Implementing the test using testground builders + runners,
- Rely on `docker:generic` and `docker_extensions` to implement missing features,
- As we converge to a reliable solution, make the features Testground natives (add a `docker:rust` builder for example, and other simplifications).
- With this approach, I would NOT recommend moving to `docker:generic`, but instead moving features into testground.

It seems we're now moving towards a different design:

- Implementing the tests using [external] builders + testground's runners.
  - related: https://github.com/testground/testground/issues/1522
- With this approach, I would recommend investing in external builders rather than rewriting what we have.